### PR TITLE
feat: ser/de for `IpAddr` type; schema for `Ipv4Addr`/`Ipv6Addr`/`IpAddr` types 

### PR DESCRIPTION
**Pull request by willco-1**
this is a WIP attempt to add the base IpAddr type to the schema.

### CODE REVIEW

File: borsh/tests/schema/test_ip_addr.rs [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+use std::net::IpAddr;
+
+#[test]
+fn ip_addr_schema() {
+    let actual_name = IpAddr::declaration();
+    let mut actual_defs = schema_map!();
+    IpAddr::add_definitions_recursively(&mut actual_defs);
+
+    assert_eq!("IpAddr", actual_name);
+    assert_eq!(
```
Comment (dj8yfo) [inline borsh/tests/schema/test_ip_addr.rs:13]: this is a suggestion of how implementation for BorshSchema  may be added https://github.com/dj8yfo/borsh-rs/commit/7ee0dc90c7ad8ca477b63c57d8054b374db25408

**Review by dj8yfo — COMMENTED** [review]
@willco-1 do you plan on finishing the pr?

**Comment by willco-1** [comment]
@dj8yfo yes! Sorry .. I got busy and forgot to come back to it. Thanks for the tips, this is very helpful.

File: borsh/tests/schema/test_ip_addr.rs [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+        schema_map! {
```
Comment (dj8yfo) [inline borsh/tests/schema/test_ip_addr.rs:14]: please, replace `schema_map!` with `insta::assert_snapshot!(format!("{:#?}", defs));` similar to how it's done here
https://github.com/near/near-sdk-rs/pull/1212/commits/bd45f91862d51816c3b1459431383d98e8086304

File: borsh/tests/schema/test_ip_addr.rs [hunk 3]
```diff
@@ -0,0 +1,35 @@
+use crate::common_macro::schema_imports::*;
+
+#[cfg(feature = "std")]
```
Comment (dj8yfo) [inline borsh/tests/schema/test_ip_addr.rs:3]: remove `#[cfg(feature = "std")]` on this line

File: borsh/tests/tests.rs [hunk 4]
```diff
@@ -73,12 +73,13 @@ mod roundtrip {
...
     #[cfg(feature = "ascii")]
-    mod test_ascii_strings; 
-    mod test_strings; 
+    mod test_ascii_strings;
+    mod test_strings;
     mod test_arrays;
     mod test_vecs;
     mod test_tuple;
     mod test_primitives;
+    mod test_ip_addr;
```
Comment (dj8yfo) [inline borsh/tests/tests.rs:82]: add `#[cfg(feature = "std")]` before this line

File: fuzz/fuzz-run/src/main.rs [hunk 5]
```diff
@@ -33,6 +33,7 @@ fn main() {
                 std::net::SocketAddrV6,
                 std::net::Ipv4Addr,
                 std::net::Ipv6Addr,
+                std::net::IpAddr,
```
Comment (dj8yfo) [inline fuzz/fuzz-run/src/main.rs:36]: revert changes to this file

File: borsh/src/ser/mod.rs [hunk 6]
```diff
@@ -505,7 +505,16 @@ impl BorshSerialize for std::net::Ipv6Addr {
     }
 }
 
-impl<T: BorshSerialize + ?Sized> BorshSerialize for Box<T> {
+#[cfg(feature = "std")]
+impl BorshSerialize for std::net::IpAddr {
+    #[inline]
+    fn serialize<W: Write>(&self, writer: &mut W) -> Result<()> {
+        match self {
+            std::net::IpAddr::V4(ipv4) => ipv4.serialize(writer),
```
Comment (dj8yfo) [inline borsh/src/ser/mod.rs:513]: this has lost tag (0u8 or 1u8) serialization before a variant's serialization

File: borsh/tests/tests.rs [hunk 7]
```diff
@@ -31,9 +31,9 @@ mod compile_derives {
...
 #[rustfmt::skip]
 mod roundtrip {
-    mod test_strings; 
+    mod test_strings;
     #[cfg(feature = "ascii")]
-    mod test_ascii_strings; 
+    mod test_ascii_strings;
     mod test_arrays;
     mod test_vecs;
     mod test_tuple;
```
Comment (dj8yfo) [inline borsh/tests/tests.rs:39]: add a module with rountrip test below line 40 (test_primitives): 
```rust
#[cfg(feature = "std")]
mod test_ip_addr;
```

it should contain 3 tests: a roundtrip test for `Ipv4Addr`, for `Ipv6Addr` and a 
a roundtrip test for `vec![IpAddr::V4{ ...}, IpAddr::V6{...}, IpAddr::V6{ ...}]`

**Comment by dj8yfo** [comment]
@race-of-sloths include

**Review by dj8yfo — APPROVED** [review]

**Pull request merged**